### ISSUE TITLE
ML随想

### ISSUE THREAD

**Issue by towerjoo** [towerjoo/myML#8]
## 2017.5.31

https://colah.github.io/posts/2015-09-Visual-Information/

colah大神的这篇关于信息论的文章非常值得一读，其中关于entropy, cross-entropy和KL divergence的解释非常生动：

1. entropy是H(p)=\sum_x{p(x)log_2(1/p(x)) 说明了熵与概率的关系
2. cross entropy: 也就是H_p(q)=\sum_x{q(x)log_2(1/p(x)) 也就是遵从(with respect to)p的概率分布（编码方式）而计算出q的熵，或者就是作者在文章中具体的编码例子，也就是alice使用bob的编码方式来传送自己信息的熵（平均长度）
3. KL divergence: D_q(p)=H_q(p)-H(p), 这就是p通过q的编码方式来传递自己信息的平均长度与p通过自己的编码方式来传递自己信息的平均长度的差，in general，KL divergence描述的是2个概率分布的距离（所谓的差别），这个在ML上非常重要，因为这是量化描述我们习得的分布与ground truth分布差别的方法，从而为改进提供了理论的量化的数学基础

![hxy-info](https://cloud.githubusercontent.com/assets/145304/26614449/10724318-45f3-11e7-95e5-80136e4e3620.png)

上面的图很好地总结了各种熵的关系，其中H(X)是X的熵，H(X|Y)是条件熵，H(X, Y)是联合熵，V(X,Y)是variable of information, I(X, Y)是共同信息(mutual information)，并且有：

```
I(X, Y)=H(X)+H(Y)-H(X,Y)
V(X, Y)=H(X,Y)-I(X,Y)
```

## 2017.5.8

from cs231n lecture 4

1. 前向传播来计算Loss，反向传播来计算gradient，然后更新参数，如此循环直到达到一定的threshold（例如error rate, loss value)
2. 理解常用门的gradient，例如add gate: 平均分布梯度, multiply gate: gradient "switch", max gate: drop(pick up)

## 2017.5.2

1. 我们应该有一个general knowledge的pre-trained model，这个pre-trained model可以被后续的各种AI task的Model来使用，而不是所有特定task的model都是从头来训练，或者类似于各个programming library，我们可以使用这些building blocks来构建我们自己特定场景下的应用，而不需要每次都reinvent wheels(自动或者手动）来重新训练这些通用知识的model
2. 从过去的研究各种问题的representation（或者更通用的knowledge representation)，到现在的end-to-end让机器自己figure out问题的representation，这是一种进步还是退步？NYU及一家Uber收购的AI公司的创始人提到：目前的统计学的方法是简单的，而研究representation是困难的，所以我们趋向于目前的end-to-end的方法。这个是有一定道理的，特别是feature enginering虽然困难，但是通常的模型是高效的（无论是训练或者evaluation阶段）

## 2017.4.27

1. end-to-end似乎成了一个流行而且主流的ML方法，优势自不用说（主要是无需手工进行feature extracting)，所有的feature extracting都是由网络自己进行的，自然使得domain knowledge不再那样重要，所有人都可以make it work，但是相应的问题来了：
    * prior knowledge(或者叫domain knowledge)是很重要的，可以将问题进行降维，使得效率自然高很多
   * 目前我们不断在增加depth的同时，似乎出现了一种想法就是足够deep的DNN可以解决所有问题，唯一的局限就是depth和hardware，这就像在说一个程序可以解决所有的问题，唯一的局限是程序的大小和hardware
    * 有人喜欢大而全，但程序员似乎更喜欢unix philosophy, i.e **do one thing, and do it well**，这里的well自然是包括性能的
2. hyperparameter的问题一直以来都是DL的一个重要的问题，一方面我们尽量减少超参数的数量，使得manual engineering越少越好，一方面我们有希望model更加*精致*，也就是更多地加入人为的设计。前者如VGGNet期望保持每一次的结构不变而只是增加深度（这样我们就可以不用考虑不同层的具体设计），而Inception又是在精致设计Model的方向不断前行。

## 2017.3.28

http://colah.github.io/posts/2014-03-NN-Manifolds-Topology/

1. 和自己一直以来思考的一样，NN的hidden layer就是representation的转化，将线性不可分的input转化为线性可分的representation，然后在output层直接draw the line（或者高维的hyperplane）来classify对象。representation是非常关键的，所以研究representation是非常重要的，例如Mikolov 2013关于words的distributed representation的论文 
2. 优于Input不是线性可分的，所以我们需要把它转化为线性可分，通常是通过非线性函数来完成，例如tanh, ReLU, sigmoid等，具体的tanh(Wx+b)
  * 通过W权重矩阵进行线性变换
  * 通过bias b进行平移
  * 通过tanh进行point wise的函数变换（非线性）
3. 进行representation变换，需要保证拓扑homeomorphisms，也就是说连接什么的都不能改变
4. 标准的NN可以看做是1）放射变换（affine transformation，也就是线性变换并且保持平直性和平行性，例如正方形变成平行四边形）2）接着point-wise的激活函数变换

## 2017.3.23

> activation function: 激活函数，该如何理解？neural networks中的每个perceptron就是一个activation function，来决定信号（特征,feature等）是否存在，例如1表示存在，0表示不存在。激活函数通常使用连续的函数而不是1，0这样的开关，是因为我们训练时希望看到参数的很小变动会对输出的影响，如果我们使用1，0则很小的变动其实不会影响输出的变动。而连续的函数则可以。另外通常激活函数是非线性的，是因为我们需要给NN引入非线性的能力，如果是线性的则整个NN也是线性的（线性之间的组合也是线性的），显然处理能力会下降。常用的激活函数有：sigmoid, ReLu, tanh等。https://www.quora.com/What-is-the-role-of-the-activation-function-in-a-neural-network提到NN是对于input space到output space的映射，是通过projection和selection来完成的，projection可以通过inner product, convolutional function等来进行，也就是参数和输入相似度的度量(x2=x1*w12, 节点1到节点2），而activation function就是用来selection的，这样可以映射到更高层次的空间（阻断一些feature）。

> 关于local minimum足够好还有一个原因是overfitting的问题，ML很有趣的一个特点是，你需要刚刚好，不能不足也不能过，但是这个good fit是很难的。

> local minimum vs. global minimum 我们学习的目标是找到全局最优解，但是当对象不是convex时，或者说有多个local minimum时，(S)GD就会困在某个local minimum(或者saddle point)处，而无法找到global minimum，初看这个似乎是个很严重的问题，因为某个local minimum可能与global minimum相差很大，但是经验（imagenet比赛的错误率）或者理论（有相关的论文）的结论是：对于复杂的网络（此处复杂网络很重要），所有的local minimum都是很接近的，与global minimum也是很接近的，所以通常找到的local minimum就足够好，而从一个local minimum到global minimum的计算是指数级的，在实践中是不可行的。 论文：http://arxiv.org/abs/1412.0233

> 当loss function不连续时，那么就不可微，导致我们无法通过凸函数的优化方法来找到最值点，这时我们通常可以使用替代函数(surrogate function)，这个surrogate function有更好的数学特征，例如凸函数（可以方便地优化）。那么问题来了，这种替换的成本是什么？http://fa.bianp.net/blog/2014/surrogate-loss-functions-in-machine-learning/ 提到了，如果替代函数是凸函数并且在0点可微而且一阶导数小于0，那么原函数与替代函数在优化上是一致的(consistent)。一个很好的例子是对于logistical classification，我们不使用阶跃函数作为f(xw) （predict function)，而使用替代函数sigmoid函数作为f(xw)，那么相应的loss function是凸函数。（这里注意区分loss function和predict function，sigmoid函数本身不是凸函数）常用的替代函数有hinge loss function(SVM), logistic loss function等。

> differentiable 是ML中非常重要的一个术语（term），这是因为基本所有的ML模型和方法最终都可转化为convex optimization（例如最小化loss function或者最大化精度)，而一个函数可微是这种优化的前提，也是梯度下降方法可用的前提（规模太大的网络通常GD是可用的唯一办法）。

## 2017.3.21

http://rylanschaeffer.github.io/content/research/overcoming_catastrophic_forgetting/main.html

> This is really interesting, since I just thought about this for some time in past days. Comparing to my 2 years old kid(3 years ago, now she's 5), she might manage to learn the cat from ~100 exposures of cats' pictures, but it will need much less pictures to learn the dog, which the reason I guess is that she have the memory(or experience, or pattern, etc) and she can apply this experience to the new object. The benefits of such memory is really cool, for 1) it requires less training data 2) it speeds up the new objects' learning.

> This paper(and the interpretation to the paper) seems promising to handle the current forgot* issue of neural network.

## 2017.3.20

> 深度学习的各种模型，确实在实际的领域中取得了非常好的进展，不过规模庞大（所谓的深）的模型，构造了一个输入和输出的黑盒子，可以工作的原因及其严格的数学论证目前仍没有好的答案，这也是我很担心的一点。没有坚实的数学最为根基，这样的大厦总让人觉得某天会倒下，哪怕此刻有多么金碧辉煌，耀眼夺目。这也是那位微软的ML专家对于openAI成立时的十人都是深度学习的应用专家，而缺少理论研究者的担心，我是很赞同这样的担心的。未来几年，除了imageNet的识别率那几个点的提高、或者dropout类似的训练tricks，我更期待在理论层面有更深入而坚实的进展。

> RNN（及相应的变种，例如LSTM），这种解决sequence-to-sequence的模型真心是很牛逼，如此的抽象可以将很多问题纳入旗下，例如问答系统、机器翻译、图片注解等等。

> 相比于人类（例如2岁小孩），深度学习效率特别低，通过CNN学习到一个猫可能需要10000+张图片，通过RNN学习3位数的加法可能需要50000条记录，而人类可能只需要100这个数量级的图片，究其原因最为重要的一点差别是人类在学习不同对象之间是有记忆的，例如如果人类首先学习了猫，那么在学习狗时，相应的特征提取的经验会自然地应用到狗的学习之中，而不是从头开始，从而大大减少了训练样本的个数及相应的训练时间。而我们的图像识别，不同的对象识别之间通常是没有记忆的，每个对象的识别都是独立的，因而需要大量的样本和训练时间。

> i.d.d（独立同分布）是机器学习研究的一个重要假设，这也是符合我们的直觉的，同一个对象（事件）是遵从相同的分布规律的，否则机器学习则成为不可能。但是分布的规律与训练集是否有关，这与统计学的2大门派的观点有关（分歧），频率学派认为规律是确定的客观的（类比于参数w是确定的），我们只需要研究样本空间来建模事件的规律（事件的角度）；而贝叶斯学派则认为事件本身具有某种随机性，从观察者的角度来看待事件的规律，在观察者只是不完备的条件下使用贝叶斯概率猜测事件规律的过程(w是不断调整的）。当然，贝叶斯派并非否定事件本身规律的客观性，而是观察者不知道而已。所以，无论是哪个学派，都是不否定规律本身的客观性，只是贝叶斯派试图去通过当前的不完备的知识来猜想（逼近）实际的规律。
